src/consensus.rs

```diff
@@ -658,15 +700,186 @@ impl Consensus {
-                self.on_commit(consensus_proposal_hash, commit_quorum_certificate)
+            ConsensusNetMessage::Commit(hash, cert) => {
+                self.on_commit(hash, cert)
+            }
+            ConsensusNetMessage::Timeout(slot, view, consensus_proposal_hash) => {
+                self.on_timeout(msg, slot, view, consensus_proposal_hash)
+            }
+            ConsensusNetMessage::TimeoutCertificate(slot, view, certificate) => {
+                self.on_timeout_certificate(view, slot, certificate)
             }
             ConsensusNetMessage::ValidatorCandidacy(candidacy) => {
                 self.on_validator_candidacy(msg, candidacy)
             }
         }
     }
+    fn broadcast_net_message(&mut self, consensus_net_message: ConsensusNetMessage) -> Result<()> {
+        // TODO: use strum to generate variant names in logs
+        _ = self
+            .bus
+            .send(OutboundMessage::broadcast(
+                self.sign_net_message(consensus_net_message)?,
+            ))
+            .context("Failed to broadcast ConsensusNetMessage msg on the bus")?;
+        Ok(())
+    }
+
+    fn send_net_message(&mut self, to: ValidatorPublicKey, consensus_net_message: ConsensusNetMessage) -> Result<()> {
+        _ = self
+            .bus
+            .send(OutboundMessage::send(
+                to,
+                self.sign_net_message(consensus_net_message)?,
+            ))
+            .context("Failed to send ConsensusNetMessage msg on the bus")?;
+        Ok(())
+    }
+    
+
+    fn on_timeout_certificate(&mut self, view: &View, slot: &Slot, certificate: &QuorumCertificate) -> Result<()> {
+        // Wrong view
+        if *view < self.bft_round_state.view {
+            bail!("Received timeout message for an obsolete view {}", view);
+        }
+
+        if *slot < self.bft_round_state.slot {
+            bail!("Received timeout message for an obsolete slot {}", slot);
+        }
+
+        self.bft_round_state
+            .timeout_certificates
+            .entry(*slot)
+            .or_default()
+            .insert(*view, certificate.clone());
+
+        self.bft_round_state.slot = *slot;
+        self.bft_round_state.view = view + 1;
+
+        if self.is_next_leader() {
+            // Send Prepare message to all validators
+            self.start_new_slot()?;
+        } else if !self.is_part_of_consensus(self.crypto.validator_pubkey()) {
+            self.send_candidacy()?;
+        }
+
+        Ok(())
+    }
+
+    fn on_timeout(&mut self, msg: &SignedWithKey<ConsensusNetMessage>, view: &View, slot: &Slot, consensus_proposal_hash: &ConsensusProposalHash) -> Result<()> {
+        // Only timeout if it is in consensus
 
+        debug!(
+            "timeout message received for slot {} and view {}",
+            slot, view
+        );
+
+        if self.is_part_of_consensus(self.crypto.validator_pubkey()) {
```

Comment:
maybe we should check also that the message came from a validator that is part of the consensus too ?

Reply:
Yes